### ISSUE TITLE
Podman may fail to mount an NFS volume if --userns=keep-id is specified.

### ISSUE THREAD

**Issue by sigint2** [containers/podman#17823]
### Issue Description

Podman may fail to mount an NFS volume if `--userns=keep-id` is specified.

Consider the following NFS volume and permissions:

NFS mounted at: 
- `/mnt/my-volume`

permissions:
 - `/mnt`, owner: `0:0`, mode: `0755`
 - `/mnt/my-volume`, owner: `1000:1000`, mode: `0700`
 - `/mnt/my-volume/data`, owner: `1000:1000`, mode: `0755`

The following command run as uid `1000` will fail:

`podman run --privileged --rm -it --userns=keep-id --volume /mnt/my-volume/data:/mnt/my-volume/data my-ctr /bin/id`

This command does not fail if any of the following are true:
- `--userns=keep-id` is not specified. 
- we pass the volume to podman as `--volume /mnt/my-volume:/mnt/my-volume`
- we use a local storage volume

I believe the reason for this is because:
- `--userns=keep-id` creates the user namespace where a high UID from `/etc/subuid` is mapped to 0 inside the container, and my uid 1000 is mapped to uid 1000 in the container (all as documented).
- podman then attempts to perform the volume mount as uid 0 in the container (the high UID on the host). this will fail, because NFS will deny the high user id from statting `/mnt/my-volume/data`, because the high UID does not have execute permissions on the parent directory (`/mnt/my-volume`).
	- `/mnt/my-volume` works because `/mnt` has world execute permissions 
	  - if we change `/mnt/my-volume` to mode `0701`, podman does not complain
   - This issue will not happen with a local storage volume (again, as expected, NFS not working with user namespaces is well documented). The high uid (that is mapped as 0) will try to access the volume and succeed, because it is root in that userns, and uid 1000 has a valid mapping in the userns. The remote NFS can't do this mapping.

I believe if podman would perform the mount as the invoking user (i.e., with the same uid mapping that the container gets when `--userns=keep-id` is not specified), then podman would succeed in mounting the volume.

Does this sound right?

### Steps to reproduce the issue

Steps to reproduce the issue
1. Create the directory tree explained in the above description, on an NFS volume.
2. Run the command in the above description.

### Describe the results you received

Podman is unable to mount a NFS directory that is owned by the invoking user.

### Describe the results you expected

Podman should be able to mount the NFS directory that is owned by the invoking user.

Workarounds involving recursive chowning are not an option in my case, since the volume is rather large.

### podman info output

```yaml
Version 4.3.0 on vanilla RHEL8.
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

N/A

### Additional information

N/A

**Comment by Luap99** [comment]
@giuseppe PTAL

**Comment by giuseppe** [comment]
could you setup the mount first?  You need to do it from the `podman unshare` environment where your user is mapped to root

**Comment by sigint2** [comment]
Would you be able to clarify how you’d like that to look? Manually bind mount from podman unshare?

**Comment by sigint2** [comment]
@giuseppe that worked, assuming you meant the below. It fails if I try to bind mount the volume to itself, so I have to use something like `/srv` as the in-between mount.

```
$ podman unshare mount --bind /mnt/my-volume/data /srv
$ podman run -v /srv:/mnt/my-volume/data --rm -it --userns=keep-id my-crt ls -ltd /mnt/my-volume/data
```

Ideally we wouldn't have to run the unshare step.

**Comment by giuseppe** [comment]
podman should not perform system-wide configuration like the one above, the user must do it.

With the step above you are creating a bind mount that will survive the container since it runs in the podman rootless mount namespace, so you'll need to umount it by yourself.

**Comment by sigint2** [comment]
I agree that podman should not perform that step. By "ideally we wouldn't have to run the unshare step", I meant that podman should be able to mount in an NFS volume if `--userns=keep-id`.

Can we re-open this please, or at least give a clearer reason for why a fix won't be implemented? 

If a fix won't be implemented, please let me know if you'll accept a documentation PR for `--userns` that clearly states `--userns=keep-id` may cause issues mounting NFS volumes. I spent a bit of time chasing this bug down and wouldn't want others to waste time in the future. Thanks.

**Comment by giuseppe** [comment]
it could probably work if the OCI runtime opens the mount from the current user namespace and pass it down to the init process where the actual bind mount happens.  Said so, it is unclear (from the OCI runtime PoV) when this should be performed and if that is what the user really expects.  This is another reason why I think it should not be done automatically.

The extra step you've done to use `podman unshare`, IMO, is the correct one so that the user is clearly asking to mount it with its ID, but it has the disadvantage that the mount must be cleaned once it is not needed anymore, so you'd need to do a `podman unshare umount $MNT_PATH`.


> If a fix won't be implemented, please let me know if you'll accept a documentation PR for `--userns` that clearly states `--userns=keep-id` may cause issues mounting NFS volumes. I spent a bit of time chasing this bug down and wouldn't want others to waste time in the future. Thanks.

yes please, that would be great.  Feel free to tag me on the PR once you've it ready and I'll be glad to review it.
